cpp/src/shuffler/postbox.cpp

```diff
@@ -31,7 +31,10 @@ std::unordered_map<ChunkID, Chunk> PostBox::extract(PartID pid) {
 std::vector<Chunk> PostBox::extract_all() {
     std::lock_guard const lock(mutex_);
     std::vector<Chunk> ret;
+
     for (auto& [_, chunks] : pigeonhole_) {
+        // reserve for the chunks size
+        ret.reserve(ret.size() + chunks.size());
```

Comment:
This is quadratic since you're not growing geometrically so you have to do `num_pigeonhole_entries * num_chunks_per_pigeonhole` allocations.